File: algorithm-design/hackerrank/sherlock_and_array/sherlockand_array.py
#!/usr/bin/env python3
# https://www.hackerrank.com/challenges/sherlock-and-array

import os
import unittest

from typing import List


def balanced_sums(arr: List[int]) -> str:
    left = [0] * len(arr)
    right = [0] * len(arr)
    for i in range(1, len(arr)):
        left[i] = left[i - 1] + arr[i - 1]
        right[len(arr) - 1 - i] = right[len(arr) - i] + arr[len(arr) - i]
    for i in range(len(arr)):
        if left[i] == right[i]:
            return 'YES'
    return 'NO'


class TestCode(unittest.TestCase):
    def runner(self, name):
        path = os.path.join(os.path.split(__file__)[0], f'input{name}.txt')
        with open(path, 'r') as handle:
            lines = handle.readlines()
            lines = [line.strip() for line in lines]
        count = int(lines[0])
        results = [0] * count
        for i in range(count):
            arguments = lines[2 + i * 2].split(' ')
            arguments = [int(item) for item in arguments]
            results[i] = balanced_sums(arguments)
        path = os.path.join(os.path.split(__file__)[0], f'output{name}.txt')
        with open(path, 'r') as handle:
            lines = handle.readlines()
            self.assertEqual(len(lines), len(results))
        for index, line in enumerate(lines):
            self.assertEqual(line.strip(), results[index])

    def test_example(self):
        self.runner('_example')

    def test_00(self):
        self.runner('00')

File: algorithms/code/hackerrank/sherlock_and_array/sherlock_and_array.rs
// https://www.hackerrank.com/challenges/sherlock-and-array

use crate::hackerrank::tester::{read_input, write_and_check_output};

pub fn balanced_sums(arr: &[i32]) -> String {
    let mut left = vec![0; arr.len()];
    let mut right = vec![0; arr.len()];
    for i in 1..arr.len() {
        left[i] = left[i - 1] + arr[i - 1];
        right[arr.len() - i - 1] = right[arr.len() - i] + arr[arr.len() - i];
    }
    for i in 0..arr.len() {
        if left[i] == right[i] {
            return "YES".to_string();
        }
    }
    "NO".to_string()
}

pub fn main_tested(name: &str) {
    let origin = file!();
    let input = read_input(origin, name);
    let t = input[0][0].parse::<u32>().unwrap();
    let mut results: Vec<String> = Vec::new();
    let mut offset = 2;
    for _ in 0..t {
        let arr: Vec<i32> = input[offset]
            .iter()
            .map(|s| s.to_string().parse::<i32>().unwrap())
            .collect();
        offset += 2;
        results.push(balanced_sums(&arr));
    }
    write_and_check_output(origin, name, &results);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_example() {
        main_tested("_example");
    }

    #[test]
    fn test00() {
        main_tested("00");
    }
}
